simSPI/crd.py

```diff
@@ -0,0 +1,135 @@
+
+
+def write_crd_file(
+    numpart,
+    xrange=np.arange(-100, 110, 10),
+    yrange=np.arange(-100, 110, 10),
+    crd_file="crd.txt",
+):
+    """Write particle data to .crd file.
+
+    Parameters
+    ----------
+    numpart : int
+        Number of particles.
+    xrange : ndarray
+        Valid horizontal range for display.
+    yrange : ndarray
+        Valid vertical range for display.
+    crd_file : str
+        Relative path to output .crd file.
+    """
+    if os.path.exists(crd_file):
+        print(crd_file + " already exists.")
+    else:
+        rotlist = get_rotlist(numpart)
+        with open(crd_file, "w") as crd:
+            crd.write("# File created by TEM-simulator, version 1.3.\n")
+            crd.write("{numpart}  6\n".format(numpart=numpart))
+            crd.write(
+                "#            \
+                x             \
+                y             \
+                z             \
+                phi           \
+                theta         \
+                psi  \n"
+            )
+            i = 0
+            for y in yrange:
+                for x in xrange:
+                    if i == int(numpart):
+                        break
+                    crd_table = {
+                        "x": x,
+                        "y": y,
+                        "z": 0,
+                        "phi": rotlist[i][0],
+                        "theta": rotlist[i][1],
+                        "psi": rotlist[i][2],
+                    }
+                    crd.write(
+                        "{0[x]:14.4f} \
+                        {0[y]:14.4f} \
+                        {0[z]:14.4f} \
+                        {0[phi]:14.4f} \
+                        {0[theta]:14.4f} \
+                        {0[psi]:14.4f}\n".format(
+                            crd_table
+                        )
+                    )
+                    i += 1
+
+
+def get_rotlist(numpart):
+    """Return a rotation list containing Euler angles.
+
+    Parameters
+    ----------
+    numpart : int
+        Number of particles.
+    """
+    rotlist = []
+    for x in range(0, numpart + 1):
+        x = special_ortho_group.rvs(3)
+        y = rotation_matrix_to_euler_angles(x)
+        rotlist.append(y)
+    return rotlist
+
+
+def rotation_matrix_to_euler_angles(R):
+    """Compute Euler angles given a rotation matrix.
+
+    Parameters
+    ----------
+    R : ndarray
+        Rotation matrix.
+    """
+    if not is_rotation_matrix(R):
+        raise ValueError()
+
+    sy = math.sqrt(R[0, 0] * R[0, 0] + R[1, 0] * R[1, 0])
+    singular = sy < 1e-6
+    if not singular:
+        x = math.atan2(R[2, 1], R[2, 2])
+        y = math.atan2(-R[2, 0], sy)
+        z = math.atan2(R[1, 0], R[0, 0])
+    else:
+        x = math.atan2(-R[1, 2], R[1, 1])
+        y = math.atan2(-R[2, 0], sy)
+        z = 0
+    x = rad2deg(x)
+    y = rad2deg(y)
+    z = rad2deg(z)
+    return np.array([x, y, z])
+
+
+def rad2deg(x):
```

Comment:
I believe this is already available in numpy: https://numpy.org/doc/stable/reference/generated/numpy.rad2deg.html

Good news! Less unit tests to implement ;)